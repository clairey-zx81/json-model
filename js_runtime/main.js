import { parseArgs } from 'node:util'
import fs from 'node:fs'

// TODO escaping?
const json_path = (path) => '.' + path.join('.')

function report(rep)
{
    if (rep === null || rep.length === 0) return ''

    let reps = []
    for (const [message, path] of rep)
        reps.push(`${json_path(path)}: ${message}`)

    return ' (' + reps.join('; ') + ')'
}

export default function main(checker)
{
    const options = {
      'verbose': { type: 'boolean', short: 'v' },
      'times': { type: 'string', short: 'T' },
      'report': { type: 'boolean', short: 'r' },
    }

    const args = parseArgs({options, allowPositionals: true})

    // args.hasOwnProperty('verbose')
    // args.verbose === undefined
    let times = args.values.times !== undefined ? Number(args.values.times) : 1

    if (isNaN(times))
    {
        console.log(`Unexpected --times: ${args.values.times}`)
        process.exit(1);
    }

    // overhead estimation
    let empty = 0.0
    if (times > 1)
    {
        let n = times
        while (n--)
        {
            const start = performance.now()
            empty += 1000.0 * (performance.now() - start)
        }
        empty /= times
    }

    for (const fname of args.positionals)
    { 
        try {
            fs.readFile(fname, (err, data) => {
                if (err) throw err
                const value = JSON.parse(data)

                // show performance
                // NOTE: jit, calls may take a lot of time at the beginning
                // NOTE: gc, may be irregular…
                if (times > 1)
                {
                    let sum = 0.0
                    let sum2 = 0.0
                    let n = times
                    while (n--) {
                        const start = performance.now()
                        let rep = args.values.report ? [] : null
                        const valid = checker(value, '', rep)
                        rep && (rep.length = 0)
                        let delay = (1000.0 * (performance.now() - start)) - empty  // µs
                        sum += delay
                        sum2 += delay * delay
                    }
                    const avg = sum / times
                    const stdev = Math.sqrt(sum2 / times - avg * avg)
                    console.log(`${fname}: ${args.values.report ? 'rep' : 'nop'}`,
                                `${avg.toFixed(3)} ± ${stdev.toFixed(3)} µs (${empty.toFixed(3)})`)
                }

                // do it anyway for the result
                let rep = args.values.report ? [] : null
                const valid = checker(value, '', rep)
                console.log(`${fname}: ${valid ? 'PASS' : 'FAIL'}${report(rep)}`)
            })
        }
        catch (e) {
            console.log(`${fname}: ERROR (${e})`)
        }
    }
}
