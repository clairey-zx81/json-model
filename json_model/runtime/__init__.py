from .types import (
    Jsonable,
    Path,
    Report,
    PropMap,
    CheckFun,
    RegexFun,
    EntryCheckFun,
)

from .support import (
    # TODO remove _
    UNDEFINED,
    _tname,
    _path,
    _rep,
    Const,
    ConstSet,
    ConstMap,
    is_valid_date,
    is_valid_time,
    is_valid_datetime,
    is_valid_regex,
    is_valid_exreg,
    is_valid_url,
    is_valid_email,
    is_valid_uuid,
    is_valid_json,
    is_unique_array,
    check_constraint,
    value_len,
    main,
)
