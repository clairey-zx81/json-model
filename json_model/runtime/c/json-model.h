#ifndef JSON_MODEL_H_INCLUDED
#define JSON_MODEL_H_INCLUDED

#define _GNU_SOURCE  // for qsort_r?
#include <stdbool.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#include <jansson.h>

/*
 * build generated API names
 */
#ifndef CHECK_FUNCTION_NAME
#warning defining CHECK_FUNCTION_NAME macro
#define CHECK_FUNCTION_NAME check_model
#endif  // !CHECK_FUNCTION_NAME

// branch hints
#define likely(x)       __builtin_expect(!!(x), 1)
#define unlikely(x)     __builtin_expect(!!(x), 0)

// inlining hints
#if defined(NO_INLINE_HINT)
#define INLINE
#else
#if defined(__clang__)
#define INLINE __attribute__((always_inline))
#elif defined(__GNUC__)
#define INLINE inline
// #elif defined(_MSC_VER)
#else
#define INLINE
#endif  // clang/gcc/...
#endif  // NO_INLINE_HINT

#define concat(prefix, suffix) prefix ## suffix
#define newname(prefix, suffix) concat(prefix, suffix)

#define CHECK_map_size newname(CHECK_FUNCTION_NAME, _map_size)
#define CHECK_map_tab newname(CHECK_FUNCTION_NAME, _map_tab)
#define CHECK_fun newname(CHECK_FUNCTION_NAME, _map)
#define CHECK_init newname(CHECK_FUNCTION_NAME, _init)
#define CHECK CHECK_FUNCTION_NAME
#define CHECK_free newname(CHECK_FUNCTION_NAME, _free)

// set on init
extern char * jm_version_string;

#define array_length(arr) (sizeof(arr) / sizeof((arr)[0]))
typedef int (*jm_cmp_fun_t)(const void *, const void *);

/*
 * JSON API extensions
 */

// keep object prop/vals as an array for sorting and fast get.
typedef struct {
    const char *prop;
    const json_t *val;
} jm_propval_t;

extern int jm_propval_cmp(const jm_propval_t *, const jm_propval_t *);

extern bool jm_json_is_scalar(const json_t *);
extern int jm_json_cmp(const json_t *, const json_t *);
extern int jm_json_array_cmp(const json_t *, const json_t *);
extern int jm_json_object_cmp(const json_t *, const json_t *);
extern bool jm_json_array_unique(const json_t *);

/*
 * reporting
 */
// linked list of report entries
typedef struct jm_report_entry {
  const char *message;
  char *path;
  struct jm_report_entry *prev;
} jm_report_entry_t;

typedef struct {
  jm_report_entry_t *entry;
} jm_report_t;

// path segment, name == NULL means use index
typedef struct jm_path {
    const char *name;
    size_t index;
    struct jm_path *prev;  // link to previous (in stack) segment
    struct jm_path *next;  // used for backtacking
} jm_path_t;

extern void jm_report_add_entry(jm_report_t* rep, const char *msg, jm_path_t *path);
extern void jm_report_free_entries(jm_report_t *rep);

typedef bool (*jm_check_fun_t)(const json_t *, jm_path_t *, jm_report_t *);

extern bool jm_check_fun_string(jm_check_fun_t, const char *, jm_path_t *, jm_report_t *);
extern bool jm_array_is_unique(const json_t *, jm_path_t *, jm_report_t *);

/*
 * property mapping management
 */
typedef struct {
    const char *name;
    jm_check_fun_t val_check;
} jm_propmap_t;

extern void jm_sort_propmap(jm_propmap_t *, int);
extern jm_check_fun_t jm_search_propmap(const char *, const jm_propmap_t *, int);

/*
 * set of scalar constants
 */
typedef enum {
    cst_is_null,
    cst_is_bool,
    cst_is_integer,
    cst_is_float,
    cst_is_string
    // cst_is_array + const json_t * + free
    // cst_is_object + const json_t * + free
} jm_constant_tag;

typedef struct
{
    jm_constant_tag tag;
    union {
        bool b;
        int64_t i;
        double f;
        const char *s;
    } val;
} jm_constant_t;

extern bool jm_set_cst(jm_constant_t *, const json_t *);
extern void jm_dbg_cst(const jm_constant_t *, const char *);
extern void jm_sort_cst(jm_constant_t *, size_t);
extern bool jm_search_cst(const jm_constant_t *, const jm_constant_t *, size_t);

/*
 * constant mapping management for discriminant optimization
 */
typedef struct {
    jm_constant_t cst;
    jm_check_fun_t check_val;
} jm_constmap_t;

extern void jm_sort_constmap(jm_constmap_t *, size_t);
extern jm_check_fun_t jm_search_constmap(const jm_constant_t *, const jm_constmap_t *, size_t);

/*
 * Miscellaneous support functions and related data
 */
extern bool jm_is_valid_date_slow(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_date_fast(const char *, jm_path_t *, jm_report_t *);
extern bool (*jm_is_valid_date)(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_time(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_datetime(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_uuid(const char *, jm_path_t *, jm_report_t *);
extern size_t jm_str_len(const char *);
extern size_t jm_any_len(json_t *);
extern bool jm_is_valid_regex_slow(const char *, bool, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_regex_fast(const char *, bool, jm_path_t *, jm_report_t *);
extern bool (*jm_is_valid_regex)(const char *, bool, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_url(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_email(const char *, jm_path_t *, jm_report_t *);
extern bool jm_is_valid_json(const char *, jm_path_t *, jm_report_t *);

typedef enum {
    op_eq,
    op_ne,
    op_le,
    op_lt,
    op_ge,
    op_gt
} jm_constraint_op_t;

extern bool
jm_check_constraint(const json_t *, jm_constraint_op_t, const jm_constant_t *,
                    jm_path_t *, jm_report_t *);

/*
 * Shared high-level entry point
 */
extern bool jm_generic_entry(const char *(*)(void), jm_check_fun_t (*)(const char *),
                             const json_t *, const char *, bool *, char **);

/*
 * Generated stuff
 */
// low-level interface
extern const size_t CHECK_map_size;
extern jm_propmap_t CHECK_map_tab[];
extern jm_check_fun_t CHECK_fun(const char *);
// high-level interface
extern const char *CHECK_init(void);
extern bool CHECK(const json_t *, const char *, bool *, char **);
extern void CHECK_free(void);

#endif  // JSON_MODEL_H_INCLUDED
