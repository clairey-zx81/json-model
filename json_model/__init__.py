#
# convenient types, classes and functions to process models
#
from .mtypes import Jsonable
from .runtime.types import Report, EntryCheckFun
from .model import JsonModel
from .script import (
    jmc_script,
    model_from_json, model_from_url, model_from_str,
    model_checker, model_checker_from_url, model_checker_from_json,
)

__ALL__ = [
    Jsonable, Report, EntryCheckFun, JsonModel,
    jmc_script,
    model_from_json, model_from_url, model_from_str,
    model_checker, model_checker_from_url, model_checker_from_json,
]
