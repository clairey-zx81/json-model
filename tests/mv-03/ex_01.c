#include <json-model.h>
#define JSON_MODEL_VERSION "2.0a0"

static bool json_model_4(const json_t *val, jm_path_t *path, jm_report_t *rep);
static bool json_model_5(const json_t *val, jm_path_t *path, jm_report_t *rep);
static bool json_model_1(const json_t *val, jm_path_t *path, jm_report_t *rep);
jm_propmap_t check_model_map_tab[3];
const size_t check_model_map_size = 3;

// check $x1 (.'$x1')
static bool json_model_4(const json_t *val, jm_path_t *path, jm_report_t *rep)
{
    bool res;
    // .'$x1'
    res = json_is_string(val) && strcmp(json_string_value(val), "Susie") == 0;
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "unexpected _Susie [.'$x1']", path);
    }
    return res;
}

// check $x2 (.'$x2')
static bool json_model_5(const json_t *val, jm_path_t *path, jm_report_t *rep)
{
    bool res;
    // .'$x2'
    res = json_is_string(val) && strcmp(json_string_value(val), "Susie") == 0;
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "unexpected _Susie [.'$x2']", path);
    }
    return res;
}

// object .
static bool _jm_obj_0(const json_t *val, jm_path_t *path, jm_report_t *rep)
{
    if (! json_is_object(val))
    {
        if (rep) jm_report_add_entry(rep, "not an object [.]", path);
        return false;
    }
    bool res;
    int64_t must_count = 0;
    const char *prop;
    json_t *pval;
    json_object_foreach((json_t *) val, prop, pval)
    {
        jm_path_t lpath_0 = (jm_path_t) { prop, 0, path, NULL };
        if (strcmp(prop, "a") == 0)
        {
            // handle must a property
            must_count += 1;
            // .a
            res = json_is_string(pval) && strcmp(json_string_value(pval), "Susie") == 0;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "unexpected _Susie [.a]", (path ? &lpath_0 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid mandatory property value [..a]", (path ? &lpath_0 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "b") == 0)
        {
            // handle must b property
            must_count += 1;
            // .b
            res = json_is_string(pval) && strcmp(json_string_value(pval), "Susie") == 0;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "unexpected _Susie [.b]", (path ? &lpath_0 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid mandatory property value [..b]", (path ? &lpath_0 : NULL));
                return false;
            }
        }
        else
        {
            if (rep) jm_report_add_entry(rep, "no other prop expected [.]", (path ? &lpath_0 : NULL));
            return false;
        }
    }
    if (must_count != 2)
    {
        if (rep != NULL)
        {
            if (! (json_object_get(val, "a") != NULL))
            {
                if (rep) jm_report_add_entry(rep, "missing mandatory prop <a> [.]", path);
            }
            if (! (json_object_get(val, "b") != NULL))
            {
                if (rep) jm_report_add_entry(rep, "missing mandatory prop <b> [.]", path);
            }
        }
        return false;
    }
    return true;
}

// check $ (.)
static bool json_model_1(const json_t *val, jm_path_t *path, jm_report_t *rep)
{
    bool res;
    // .
    res = _jm_obj_0(val, path, rep);
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "not an expected object [.]", path);
    }
    return res;
}

jm_check_fun_t check_model_map(const char *pname)
{
    return jm_search_propmap(pname, check_model_map_tab, 3);
}

static bool initialized = false;

char *CHECK_init(void)
{
    if (!initialized)
    {
        initialized = true;
        jm_version_string = JSON_MODEL_VERSION;
        check_model_map_tab[0] = (jm_propmap_t) { "", json_model_1 };
        check_model_map_tab[1] = (jm_propmap_t) { "x1", json_model_4 };
        check_model_map_tab[2] = (jm_propmap_t) { "x2", json_model_5 };
        jm_sort_propmap(check_model_map_tab, 3);
    }
    return NULL;
}

void CHECK_free(void)
{
    if (initialized)
    {
        initialized = false;

        // cleanup code
    }
}


/*
 * # API
 *
 * ```c
 * bool error;
 * char *message;
 *
 * bool valid = check(json_value, "model-name", &error, &message);
 *
 * if (message)
 *    free(message), message = NULL;
 * ```
 *
 * - `error` is set of there was an error during initialization or if the model does not exists.
 * - `message` is provided if a non NULL pointer is passed, and must be freed by the caller.
 */
bool
CHECK(const json_t *val, const char *name, bool *error, char **reasons)
{
    return jm_generic_entry(CHECK_init, CHECK_fun, val, name, error, reasons);
}
