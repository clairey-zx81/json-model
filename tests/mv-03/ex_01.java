//
// Generated by JSON Model Compiler version 2.0b1
// see https://github.com/clairey-zx81/json-model
//
import json_model.*;
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class ex_01 extends ModelChecker
{
    static public final String VERSION = "2.0b1";

    public Map<String, Checker> ex_01_map_pmap;

    // check $x1 (.'$x1')
    public boolean json_model_4(Object val, Path path, Report rep)
    {
        boolean res;
        // .'$x1'
        res = json.isString(val) && json.asString(val).compareTo("Susie") == 0;
        if (! res)
        {
            if (rep != null) rep.addEntry("unexpected _Susie [.'$x1']", path);
        }
        return res;
    }
    
    // check $x2 (.'$x2')
    public boolean json_model_5(Object val, Path path, Report rep)
    {
        boolean res;
        // .'$x2'
        res = json.isString(val) && json.asString(val).compareTo("Susie") == 0;
        if (! res)
        {
            if (rep != null) rep.addEntry("unexpected _Susie [.'$x2']", path);
        }
        return res;
    }
    
    // object .
    public boolean _jm_obj_0(Object val, Path path, Report rep)
    {
        if (! json.isObject(val))
        {
            if (rep != null) rep.addEntry("not an object [.]", path);
            return false;
        }
        boolean res;
        long must_count = 0;
        Iterator<String> prop_loop = json.objectIterator(val);
        while (prop_loop.hasNext())
        {
            String prop = prop_loop.next();
            Object pval = json.objectValue(val, prop);
            Path lpath_0 = new Path(prop, path);
            if (prop.compareTo("a") == 0)
            {
                // handle must a property
                must_count += 1;
                // .a
                res = json.isString(pval) && json.asString(pval).compareTo("Susie") == 0;
                if (! res)
                {
                    if (rep != null) rep.addEntry("unexpected _Susie [.a]", (path != null ? lpath_0 : null));
                }
                if (! res)
                {
                    if (rep != null) rep.addEntry("invalid mandatory prop value [.a]", (path != null ? lpath_0 : null));
                    return false;
                }
            }
            else if (prop.compareTo("b") == 0)
            {
                // handle must b property
                must_count += 1;
                // .b
                res = json.isString(pval) && json.asString(pval).compareTo("Susie") == 0;
                if (! res)
                {
                    if (rep != null) rep.addEntry("unexpected _Susie [.b]", (path != null ? lpath_0 : null));
                }
                if (! res)
                {
                    if (rep != null) rep.addEntry("invalid mandatory prop value [.b]", (path != null ? lpath_0 : null));
                    return false;
                }
            }
            else
            {
                if (rep != null) rep.addEntry("unexpected prop [.]", (path != null ? lpath_0 : null));
                return false;
            }
        }
        if (must_count != 2)
        {
            if (rep != null)
            {
                if (! json.objectHasProp(val, "a"))
                {
                    if (rep != null) rep.addEntry("missing mandatory prop <a> [.]", path);
                }
                if (! json.objectHasProp(val, "b"))
                {
                    if (rep != null) rep.addEntry("missing mandatory prop <b> [.]", path);
                }
            }
            return false;
        }
        return true;
    }
    
    // check $ (.)
    public boolean json_model_1(Object val, Path path, Report rep)
    {
        boolean res;
        // .
        res = _jm_obj_0(val, path, rep);
        if (! res)
        {
            if (rep != null) rep.addEntry("unexpected element [.]", path);
        }
        return res;
    }
    

    public void init(JSON json)
    {
        if (!initialized)
        {
            try {
            ex_01_map_pmap = new HashMap<String, Checker>();
            ex_01_map_pmap.put("", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_1(o, p, r);} });
            ex_01_map_pmap.put("x1", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_4(o, p, r);} });
            ex_01_map_pmap.put("x2", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_5(o, p, r);} });
                super.init(json);
            }
            catch (Exception e) {
                throw new Error("cannot initialized model checker: " + e);
            }
        }
    }

    public void free()
    {
        if (initialized)
        {
            super.free();
            ex_01_map_pmap = null;
        }
    }

    public Checker get(String name)
    {
        return ex_01_map_pmap.get(name);
    }

    static public void main(String[] args) throws Exception
    {
        ModelChecker checker = new ex_01();
        Main.main("ex_01", checker, VERSION, args);
    }
}
