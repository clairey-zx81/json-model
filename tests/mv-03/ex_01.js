#! /bin/env node
// we may need require to load re2
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);
//
// Generated by JSON Model Compiler version 2.0b1
// see https://github.com/clairey-zx81/json-model
//
import * as runtime from "json_model_runtime"
const JSON_MODEL_VERSION = "2";

var check_model_map = new Map()

// check $x1 (.'$x1')
function json_model_4(val, path, rep)
{
    let res;
    // .'$x1'
    res = ((typeof val === 'string' || val instanceof String)) && val == "Susie";
    if (! res)
    {
        rep !== null && rep.push(["unexpected _Susie [.'$x1']", path])
    }
    return res;
}

// check $x2 (.'$x2')
function json_model_5(val, path, rep)
{
    let res;
    // .'$x2'
    res = ((typeof val === 'string' || val instanceof String)) && val == "Susie";
    if (! res)
    {
        rep !== null && rep.push(["unexpected _Susie [.'$x2']", path])
    }
    return res;
}

// object .
function _jm_obj_0(val, path, rep)
{
    if (! (Object.prototype.toString.call(val) === '[object Object]'))
    {
        rep !== null && rep.push(["not an object [.]", path])
        return false;
    }
    let res;
    let must_count = 0;
    for (const [prop, pval] of Object.entries(val))
    {
        let lpath_0 = path ? path.concat([prop]) : null;
        if (prop == "a")
        {
            // handle must a property
            must_count += 1;
            // .a
            res = ((typeof pval === 'string' || pval instanceof String)) && pval == "Susie";
            if (! res)
            {
                rep !== null && rep.push(["unexpected _Susie [.a]", (path ? lpath_0 : null)])
            }
            if (! res)
            {
                rep !== null && rep.push(["invalid mandatory prop value [.a]", (path ? lpath_0 : null)])
                return false;
            }
        }
        else if (prop == "b")
        {
            // handle must b property
            must_count += 1;
            // .b
            res = ((typeof pval === 'string' || pval instanceof String)) && pval == "Susie";
            if (! res)
            {
                rep !== null && rep.push(["unexpected _Susie [.b]", (path ? lpath_0 : null)])
            }
            if (! res)
            {
                rep !== null && rep.push(["invalid mandatory prop value [.b]", (path ? lpath_0 : null)])
                return false;
            }
        }
        else
        {
            rep !== null && rep.push(["unexpected prop [.]", (path ? lpath_0 : null)])
            return false;
        }
    }
    if (must_count != 2)
    {
        if (rep !== null)
        {
            if (! val.hasOwnProperty("a"))
            {
                rep !== null && rep.push(["missing mandatory prop <a> [.]", path])
            }
            if (! val.hasOwnProperty("b"))
            {
                rep !== null && rep.push(["missing mandatory prop <b> [.]", path])
            }
        }
        return false;
    }
    return true;
}

// check $ (.)
function json_model_1(val, path, rep)
{
    let res;
    // .
    res = _jm_obj_0(val, path, rep);
    if (! res)
    {
        rep !== null && rep.push(["unexpected element [.]", path])
    }
    return res;
}


var initialized = false

// differed module initializations
export function check_model_init()
{
    if (! initialized)
    {
        initialized = true;
        runtime.jm_set_rx(RegExp)
        check_model_map.set("", json_model_1)
        check_model_map.set("x1", json_model_4)
        check_model_map.set("x2", json_model_5)
    }
}

// differed module cleanup
export function check_model_free()
{
    if (initialized)
    {
        initialized = false;
    }
}

export function check_model(val, name, rep)
{
    let checker = check_model_map.get(name)
    if (checker === undefined)
        throw `no checker for "${name}"`
    let path = rep !== null ? [] : null

    return checker(val, path, rep)
}

// possibly run as main based on a guess
import main from "json_model_runtime/main.js"

if (import.meta.url.endsWith(process.argv[1]))
    main(check_model_init, check_model, check_model_free)
