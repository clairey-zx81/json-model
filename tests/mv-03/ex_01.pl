#! /usr/bin/perl
#
# Generated by JSON Model Compiler version 2.0b1
# see https://github.com/clairey-zx81/json-model
#
use strict;
use warnings;
use re::engine::RE2;
use JsonModel;
use constant JMC_VERSION => '2';


my %check_model_map;

# check $x1 (.'$x1')
sub json_model_4($$$)
{
    my ($val, $path, $rep) = @_;
    my $res;
    # .'$x1'
    $res = jm_is_string($val) && $val eq 'Susie';
    return $res;
}

# check $x2 (.'$x2')
sub json_model_5($$$)
{
    my ($val, $path, $rep) = @_;
    my $res;
    # .'$x2'
    $res = jm_is_string($val) && $val eq 'Susie';
    return $res;
}

# object .
sub _jm_obj_0($$$)
{
    my ($val, $path, $rep) = @_;
    if (! jm_is_object($val))
    {
        return 0;
    }
    my $res;
    my $must_count = 0;
    scalar keys %$val;
    while (my ($prop, $pval) = each %$val)
    {
        if ($prop eq 'a')
        {
            # handle must a property
            $must_count++;
            # .a
            $res = jm_is_string($pval) && $pval eq 'Susie';
            if (! $res)
            {
                return 0;
            }
        }
        elsif ($prop eq 'b')
        {
            # handle must b property
            $must_count++;
            # .b
            $res = jm_is_string($pval) && $pval eq 'Susie';
            if (! $res)
            {
                return 0;
            }
        }
        else
        {
            return 0;
        }
    }
    if ($must_count != 2)
    {
        return 0;
    }
    return 1;
}

# check $ (.)
sub json_model_1($$$)
{
    my ($val, $path, $rep) = @_;
    my $res;
    # .
    $res = _jm_obj_0($val, $path, $rep);
    return $res;
}


# initialization of global variables

our $initialized = 0;

sub check_model_init()
{
    if (!$initialized)
    {
        $initialized = 1;
        %check_model_map = (
            '' => \&json_model_1,
            'x1' => \&json_model_4,
            'x2' => \&json_model_5,
        );
    }
}

sub check_model_free()
{
    if ($initialized)
    {
        $initialized = 0;
        %check_model_map = ();
    }
}

sub check_model($$$)
{
    my ($json, $name, $rep) = @_;
    die "unexpected model name \"$name\"" unless exists $check_model_map{$name};
    my $path = defined $rep ? [] : undef;
    return $check_model_map{$name}($json, $path, $rep);
}

#
# SCRIPT EXECUTION
#
check_model_init();
jm_main(\&check_model, \%check_model_map, JMC_VERSION);
check_model_free();
