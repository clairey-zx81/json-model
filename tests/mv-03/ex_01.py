#! /bin/env python
#
# Generated by JSON Model Compiler version 2.0a0
# see https://github.com/clairey-zx81/json-model
#
from typing import Callable
import re2 as re
from json_model.runtime import *
__version__ = "2.0a0"

def check_model_fun(name: str) -> CheckFun:
    """Return check function for JSON model name."""
    return check_model_map[name]

# entry point for generated checkers
def check_model(val: Jsonable, name: str = "", rep: Report = None) -> bool:
    """Check val validity agains JSON Model name."""
    if name not in check_model_map:
        raise Exception(f"unexpected model name: {name}")
    checker = check_model_map[name]
    return checker(val, [], rep)

check_model_map: PropMap

# check $x1 (.'$x1')
def json_model_4(val: Jsonable, path: Path, rep: Report) -> bool:
    res: bool
    # .'$x1'
    res = isinstance(val, str) and val == "Susie"
    if not res:
        rep is None or rep.append(("unexpected _Susie [.'$x1']", path))
    return res

# check $x2 (.'$x2')
def json_model_5(val: Jsonable, path: Path, rep: Report) -> bool:
    res: bool
    # .'$x2'
    res = isinstance(val, str) and val == "Susie"
    if not res:
        rep is None or rep.append(("unexpected _Susie [.'$x2']", path))
    return res

# object .
def _jm_obj_0(val: Jsonable, path: Path, rep: Report) -> bool:
    if not isinstance(val, dict):
        rep is None or rep.append(("not an object [.]", path))
        return False
    res: bool
    must_count: int = 0
    for prop, pval in val.items():
        assert isinstance(prop, str)
        lpath_0: Path = (path + [ prop ]) if path is not None else None
        if prop == "a":
            # handle must a property
            must_count += 1
            # .a
            res = isinstance(pval, str) and pval == "Susie"
            if not res:
                rep is None or rep.append(("unexpected _Susie [.a]", lpath_0 if path is not None else None))
            if not res:
                rep is None or rep.append(("invalid mandatory property value [..a]", lpath_0 if path is not None else None))
                return False
        elif prop == "b":
            # handle must b property
            must_count += 1
            # .b
            res = isinstance(pval, str) and pval == "Susie"
            if not res:
                rep is None or rep.append(("unexpected _Susie [.b]", lpath_0 if path is not None else None))
            if not res:
                rep is None or rep.append(("invalid mandatory property value [..b]", lpath_0 if path is not None else None))
                return False
        else:
            rep is None or rep.append(("no other prop expected [.]", lpath_0 if path is not None else None))
            return False
    if must_count != 2:
        if rep is not None:
            if not "a" in val:
                rep is None or rep.append(("missing mandatory prop <a> [.]", path))
            if not "b" in val:
                rep is None or rep.append(("missing mandatory prop <b> [.]", path))
        return False
    return True

# check $ (.)
def json_model_1(val: Jsonable, path: Path, rep: Report) -> bool:
    res: bool
    # .
    res = _jm_obj_0(val, path, rep)
    if not res:
        rep is None or rep.append(("not an expected object [.]", path))
    return res


# initialization guard
initialized: bool = False

# differed module initializations
def check_model_init():
    global initialized
    if not initialized:
        initialized = True
        global check_model_map
        check_model_map = {
            "": json_model_1,
            "x1": json_model_4,
            "x2": json_model_5,
        }

# differed module cleanup
def check_model_free():
    global initialized
    if initialized:
        initialized = False


if __name__ == "__main__":
    check_model_init()
    main(check_model_fun, check_model_map, __version__)
    check_model_free()
