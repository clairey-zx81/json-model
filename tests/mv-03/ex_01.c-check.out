ex_01.values.json[0]: PASS
ex_01.values.json[1]: FAIL (.: not an expected object [.]; .a: invalid mandatory property value [..a]; .a: unexpected _Susie [.a])
ex_01.values.json[2]: FAIL (.: not an expected object [.]; .: missing mandatory prop <b> [.])
ex_01.values.json[3]: FAIL (.: not an expected object [.]; .: not an object [.])
ex_01.values.json[4]: FAIL (.: not an expected object [.]; .: not an object [.])
